<a href="https://colab.research.google.com/github/tashramsden/100_days_python/blob/main/Day_75_plotly_app_store.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Summary:**

* Pull a random sample from a DataFrame using .sample()

* How to find duplicate entries with .duplicated() and .drop_duplicates()

* How to convert string and object data types into numbers with .to_numeric()

* How to use plotly to generate beautiful pie, donut, and bar charts as well as box and scatter plots

# Introduction

In this notebook, we will do a comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store.

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [ ]:
import pandas as pd
import plotly.express as px

# Notebook Presentation

In [ ]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [ ]:
df_apps = pd.read_csv('apps.csv')

# Data Cleaning

**Challenge**: How many rows and columns does `df_apps` have? What are the column names? Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).

In [ ]:
df_apps.head()  # lots of nan, $ symbols in price column

App  ...         Android_Ver
0                       Ak Parti Yardım Toplama  ...          4.1 and up
1                    Ain Arabic Kids Alif Ba ta  ...          3.0 and up
2  Popsicle Launcher for Android P 9.0 launcher  ...          4.2 and up
3                     Command & Conquer: Rivals  ...  Varies with device
4                                    CX Network  ...          4.1 and up

[5 rows x 12 columns]

In [ ]:
df_apps.shape

(10841, 12)

In [ ]:
df_apps.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size_MBs', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Android_Ver'],
      dtype='object')

In [ ]:
df_apps.sample(5)

App  ...         Android_Ver
10103                                     Kick the Buddy  ...          4.4 and up
4697   Digital Clock : Simple, Tiny, Ad-free Desk Clock.  ...          2.3 and up
5920                                     Super Dancer VN  ...        4.0.3 and up
10742                       Gboard - the Google Keyboard  ...  Varies with device
10722              Google Duo - High Quality Video Calls  ...          4.4 and up

[5 rows x 12 columns]

### Drop Unused Columns

**Challenge**: Remove the columns called `Last_Updated` and `Android_Version` from the DataFrame. We will not use these columns. 

In [ ]:
df_apps.drop(columns=["Last_Updated", "Android_Ver"], inplace=True)
df_apps.head()

App  ...           Genres
0                       Ak Parti Yardım Toplama  ...           Social
1                    Ain Arabic Kids Alif Ba ta  ...        Education
2  Popsicle Launcher for Android P 9.0 launcher  ...  Personalization
3                     Command & Conquer: Rivals  ...         Strategy
4                                    CX Network  ...         Business

[5 rows x 10 columns]

### Find and Remove NaN values in Ratings

**Challenge**: How may rows have a NaN value (not-a-number) in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows. 

In [ ]:
df_apps["Rating"].isna().values.sum()

1474

In [ ]:
df_apps_clean = df_apps.dropna()
df_apps_clean.shape

(9367, 10)

### Find and Remove Duplicates

**Challenge**: Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`. 


In [ ]:
df_apps_clean["App"].duplicated().sum()  # lots of duplicates

1170

In [ ]:
df_apps_clean[df_apps_clean["App"] == "Instagram"]  # 4 duplicates for instagram

App Category  Rating   Reviews  ...  Type Price Content_Rating  Genres
10806  Instagram   SOCIAL    4.50  66577313  ...  Free     0           Teen  Social
10808  Instagram   SOCIAL    4.50  66577446  ...  Free     0           Teen  Social
10809  Instagram   SOCIAL    4.50  66577313  ...  Free     0           Teen  Social
10810  Instagram   SOCIAL    4.50  66509917  ...  Free     0           Teen  Social

[4 rows x 10 columns]

In [ ]:
df_apps_clean = df_apps_clean.drop_duplicates(subset=["App", "Type", "Price"])

In [ ]:
df_apps_clean.shape

(8199, 10)

So we can see that 13 different features were originally scraped from the Google Play Store.

* Obviously, the data is just a sample out of all the Android apps. It doesn't include all Android apps of which there are millions.

* I’ll assume that the sample is representative of the App Store as a whole. This is not necessarily the case as, during the web scraping process, this sample was served up based on geographical location and user behaviour of the person who scraped it - in our case Lavanya Gupta.

* The data was compiled around 2017/2018. The pricing data reflect the price in USD Dollars at the time of scraping. (developers can offer promotions and change their app’s pricing).

* I’ve converted the app’s size to a floating-point number in MBs. If data was missing, it has been replaced by the average size for that category.

* The installs are not the exact number of installs. If an app has 245,239 installs then Google will simply report an order of magnitude like 100,000+. I’ve removed the '+' and we’ll assume the exact number of installs in that column for simplicity.

# Find Highest Rated Apps

**Challenge**: Identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

Some highly rated apps have only a few (or even 1!) review.

In [ ]:
df_apps_clean.sort_values("Rating", ascending=False).head()

App     Category  ...  Content_Rating       Genres
21    KBA-EZ Health Guide      MEDICAL  ...        Everyone      Medical
1230         Sway Medical      MEDICAL  ...        Everyone      Medical
1227    AJ Men's Grooming    LIFESTYLE  ...        Everyone    Lifestyle
1224       FK Dedinje BGD       SPORTS  ...        Everyone       Sports
1223      CB VIDEO VISION  PHOTOGRAPHY  ...        Everyone  Photography

[5 rows x 10 columns]

In [ ]:
df_apps_clean[df_apps_clean["Rating"] == 5]  # 271 apps with rating of 5 (max)

App  ...         Genres
21                                 KBA-EZ Health Guide  ...        Medical
28                                            Ra Ga Ba  ...         Arcade
47                                             Mu.F.O.  ...         Arcade
82                                    Brick Breaker BR  ...         Arcade
99     Anatomy & Physiology Vocabulary Exam Review App  ...        Medical
...                                                ...  ...            ...
2680                               Florida Wildflowers  ...      Education
2750  Superheroes, Marvel, DC, Comics, TV, Movies News  ...         Comics
3030           CL Keyboard - Myanmar Keyboard (No Ads)  ...          Tools
3115                                        Oración CX  ...      Lifestyle
4058                          Ek Bander Ne Kholi Dukan  ...  Entertainment

[271 rows x 10 columns]

# Find 5 Largest Apps in terms of Size (MBs)

**Challenge**: What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. Based on the data, do you think there could be limit in place or can developers make apps as large as they please? 

In [ ]:
df_apps_clean.sort_values("Size_MBs", ascending=False).head()

App  ...                  Genres
9942   Talking Babsy Baby: Baby Games  ...  Lifestyle;Pretend Play
10687          Hungry Shark Evolution  ...                  Arcade
9943            Miami crime simulator  ...                  Action
9944     Gangster Town: Vice District  ...              Simulation
3144                       Vi Trainer  ...        Health & Fitness

[5 rows x 10 columns]

In [ ]:
df_apps_clean[df_apps_clean["Size_MBs"] == 100]

App  ...                  Genres
1795                               Navi Radiography Pro  ...                 Medical
3144                                         Vi Trainer  ...        Health & Fitness
4176   Car Crash III Beam DH Real Damage Simulator 2018  ...                  Racing
7926                                          Post Bank  ...                 Finance
7927                        The Walking Dead: Our World  ...                  Action
7928                      Stickman Legends: Shadow Wars  ...                  Action
8718                  Mini Golf King - Multiplayer Game  ...                  Sports
8719                         Draft Simulator for FUT 18  ...                  Sports
9942                     Talking Babsy Baby: Baby Games  ...  Lifestyle;Pretend Play
9943                              Miami crime simulator  ...                  Action
9944                       Gangster Town: Vice District  ...              Simulation
9945                                    Ultimate Tennis  ...                  Sports
10295                                   SimCity BuildIt  ...              Simulation
10687                            Hungry Shark Evolution  ...                  Arcade

[14 rows x 10 columns]

Here we can clearly see that there seems to be an upper bound of 100 MB for the size of an app. A quick google search would also have revealed that this limit is imposed by the Google Play Store itself. It’s interesting to see that a number of apps actually hit that limit exactly.

# Find the 5 App with Most Reviews

**Challenge**: Which apps have the highest number of reviews? Are there any paid apps among the top 50?

In [ ]:
df_apps_clean.sort_values("Reviews", ascending=False).head(50)

App  ...                     Genres
10805                                           Facebook  ...                     Social
10785                                 WhatsApp Messenger  ...              Communication
10806                                          Instagram  ...                     Social
10784           Messenger – Text and Video Chat for Free  ...              Communication
10650                                     Clash of Clans  ...                   Strategy
10744            Clean Master- Space Cleaner & Antivirus  ...                      Tools
10835                                     Subway Surfers  ...                     Arcade
10828                                            YouTube  ...    Video Players & Editors
10746  Security Master - Antivirus, VPN, AppLock, Boo...  ...                      Tools
10584                                       Clash Royale  ...                   Strategy
10763                                   Candy Crush Saga  ...                     Casual
10770        UC Browser - Fast Download Private & Secure  ...              Communication
10735                                           Snapchat  ...                     Social
10489    360 Security - Free Antivirus, Booster, Cleaner  ...                      Tools
10731                                     My Talking Tom  ...                     Casual
10594                                        8 Ball Pool  ...                     Sports
10302  DU Battery Saver - Battery Charger & Battery Life  ...                      Tools
10354                        BBM - Free Calls & Messages  ...              Communication
10549  Cache Cleaner-DU Speed Booster (booster & clea...  ...                      Tools
10757                                            Twitter  ...           News & Magazines
10721                                    Viber Messenger  ...              Communication
10578                                     Shadow Fight 2  ...                     Action
10813                                      Google Photos  ...                Photography
10724                        LINE: Free Calls & Messages  ...              Communication
10717                                                Pou  ...                     Casual
10792                      Skype - free IM & video calls  ...              Communication
10628                                         Pokémon GO  ...                  Adventure
10388           Minion Rush: Despicable Me Official Game  ...  Casual;Action & Adventure
10694                                            Yes day  ...                     Casual
10695                                            Hay Day  ...                     Casual
10644                           Dream League Soccer 2018  ...                     Sports
10696                                  My Talking Angela  ...                     Casual
10660             VivaVideo - Video Editor & Photo Movie  ...    Video Players & Editors
10786                       Google Chrome: Fast & Secure  ...              Communication
10817                          Maps - Navigate & Explore  ...             Travel & Local
10672                                  Hill Climb Racing  ...                     Racing
10734                                      Facebook Lite  ...                     Social
10649                                Asphalt 8: Airborne  ...                     Racing
10699                          Mobile Legends: Bang Bang  ...                     Action
10322  Battery Doctor-Battery Life Saver & Battery Co...  ...                      Tools
10396                                     Piano Tiles 2™  ...                     Arcade
10777                                       Temple Run 2  ...                     Action
10822                                             Google  ...                      Tools
10359  Truecaller: Caller ID, SMS spam blocking & Dialer  ...              Communication
10711                         SHAREit - Transfer

In [ ]:
most_reviews = df_apps_clean.sort_values("Reviews", ascending=False)[:50]
most_reviews[most_reviews["Price"] != "0"]

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size_MBs, Installs, Type, Price, Content_Rating, Genres]
Index: []

None of the top 50 apps with the most reviews are paid apps.

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

In [ ]:
ratings = df_apps_clean.Content_Rating.value_counts()
ratings

Everyone           6621
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: Content_Rating, dtype: int64

To create a pie chart we simply call px.pie() and then .show() the resulting figure. Plotly refers to all their figures, be they line charts, bar charts, or pie charts as graph_objects.

In [ ]:
fig = px.pie(labels = ratings.index, values = ratings.values)
fig.show()

Let’s customise our pie chart. Looking at the [.pie() documentation](https://plotly.com/python-api-reference/generated/plotly.express.pie.html) we see a number of parameters that we can set, like title or names.

In [ ]:
fig = px.pie(labels = ratings.index, values = ratings.values,
             title="Content Ratings", names=ratings.index)

fig.show()

If you’d like to configure other aspects of the chart, that you can’t see in the list of parameters, you can call a method called .update_traces(). In plotly lingo, “traces” refer to graphical marks on a figure. Think of “traces” as collections of attributes. Here we update the traces to change how the text is displayed.

In [ ]:
fig = px.pie(labels = ratings.index, values = ratings.values,
             title="Content Ratings", names=ratings.index)

fig.update_traces(textposition='outside', textinfo='percent+label')

fig.show()

To create a donut 🍩 chart, we can simply add a value for the hole argument:

In [ ]:
fig = px.pie(labels = ratings.index, values = ratings.values,
             title="Content Ratings", names=ratings.index,
             hole=0.6)

fig.update_traces(textposition='inside', textinfo='percent', textfont_size=15)

fig.show()

# Numeric Type Conversion: Examine the Number of Installs

**Challenge**: How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install? 

Check the datatype of the Installs column.

Count the number of apps at each level of installations. 

Convert the number of installations (the Installs column) to a numeric data type. Hint: this is a 2-step process. You'll have make sure you remove non-numeric characters first. 

In [ ]:
df_apps_clean["Installs"].describe()  # dtype: object

count          8199
unique           19
top       1,000,000
freq           1417
Name: Installs, dtype: object

In [ ]:
df_apps_clean.info()  # also tells about data types of columns

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8199 entries, 21 to 10835
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8199 non-null   object 
 1   Category        8199 non-null   object 
 2   Rating          8199 non-null   float64
 3   Reviews         8199 non-null   int64  
 4   Size_MBs        8199 non-null   float64
 5   Installs        8199 non-null   object 
 6   Type            8199 non-null   object 
 7   Price           8199 non-null   object 
 8   Content_Rating  8199 non-null   object 
 9   Genres          8199 non-null   object 
dtypes: float64(2), int64(1), object(7)
memory usage: 704.6+ KB


In [ ]:
df_apps_clean["Installs"].value_counts()

1,000,000        1417
100,000          1096
10,000            988
10,000,000        933
1,000             698
5,000,000         607
500,000           504
50,000            457
5,000             425
100               303
50,000,000        202
500               199
100,000,000       189
10                 69
50                 56
500,000,000        24
1,000,000,000      20
5                   9
1                   3
Name: Installs, dtype: int64

In [ ]:
df_apps_clean["Installs"] = df_apps_clean["Installs"].astype(str).str.replace(",", "")
df_apps_clean["Installs"] = pd.to_numeric(df_apps_clean["Installs"])
df_apps_clean[['App', 'Installs']].groupby('Installs').count()

App
Installs        
1              3
5              9
10            69
50            56
100          303
500          199
1000         698
5000         425
10000        988
50000        457
100000      1096
500000       504
1000000     1417
5000000      607
10000000     933
50000000     202
100000000    189
500000000     24
1000000000    20

Only 3 apps with a single install.

20 apps with more than A BILLION installs.

In [ ]:
df_apps_clean.sort_values("Installs", ascending=False).head(20)

App  ...                   Genres
10835                            Subway Surfers  ...                   Arcade
10806                                 Instagram  ...                   Social
10783                         Google Play Books  ...        Books & Reference
10784  Messenger – Text and Video Chat for Free  ...            Communication
10785                        WhatsApp Messenger  ...            Communication
10786              Google Chrome: Fast & Secure  ...            Communication
10788                                  Hangouts  ...            Communication
10792             Skype - free IM & video calls  ...            Communication
10803                         Google Play Games  ...            Entertainment
10805                                  Facebook  ...                   Social
10787                                     Gmail  ...            Communication
10807                                   Google+  ...                   Social
10817                 Maps - Navigate & Explore  ...           Travel & Local
10818                        Google Street View  ...           Travel & Local
10822                                    Google  ...                    Tools
10824                              Google Drive  ...             Productivity
10828                                   YouTube  ...  Video Players & Editors
10829                   Google Play Movies & TV  ...  Video Players & Editors
10831                               Google News  ...         News & Magazines
10813                             Google Photos  ...              Photography

[20 rows x 10 columns]

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

**Challenge**: Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.




In [ ]:
df_apps_clean["Price"] = df_apps_clean["Price"].astype(str).str.replace("$", "")
df_apps_clean["Price"] = pd.to_numeric(df_apps_clean["Price"])

In [ ]:
df_apps_clean.sort_values('Price', ascending=False).head(20)
# 15 or so apps that cost > $250 which are all the same - dubious, probably manipulated

App   Category  ...  Content_Rating         Genres
3946        I'm Rich - Trump Edition  LIFESTYLE  ...        Everyone      Lifestyle
2461              I AM RICH PRO PLUS    FINANCE  ...        Everyone        Finance
4606               I Am Rich Premium    FINANCE  ...        Everyone        Finance
3145              I am rich(premium)    FINANCE  ...        Everyone        Finance
3554                      💎 I'm rich  LIFESTYLE  ...        Everyone      Lifestyle
5765                       I am rich  LIFESTYLE  ...        Everyone      Lifestyle
1946  I am rich (Most expensive app)    FINANCE  ...            Teen        Finance
2775                   I Am Rich Pro     FAMILY  ...        Everyone  Entertainment
3221                  I am Rich Plus     FAMILY  ...        Everyone  Entertainment
3114                       I am Rich    FINANCE  ...        Everyone        Finance
1331          most expensive app (H)     FAMILY  ...        Everyone  Entertainment
2394                      I am Rich!    FINANCE  ...        Everyone        Finance
3897                       I Am Rich     FAMILY  ...        Everyone  Entertainment
2193             I am extremely Rich  LIFESTYLE  ...        Everyone      Lifestyle
3856                   I am rich VIP  LIFESTYLE  ...        Everyone      Lifestyle
2281       Vargo Anesthesia Mega App    MEDICAL  ...        Everyone        Medical
1407                    LTC AS Legal    MEDICAL  ...        Everyone        Medical
2629                I am Rich Person  LIFESTYLE  ...        Everyone      Lifestyle
2481         A Manual of Acupuncture    MEDICAL  ...        Everyone        Medical
4264         Golfshot Plus: Golf GPS     SPORTS  ...        Everyone         Sports

[20 rows x 10 columns]

What’s going on here? There are 15 I am Rich Apps in the Google Play Store apparently. They all cost $300 or more, which is the main point of the app. The story goes that in 2008, Armin Heinrich released the very first I am Rich app in the iOS App Store for 999.90 dollars. The app does absolutely nothing. It just displays the picture of a gemstone and can be used to prove to your friends how rich you are. Armin actually made a total of 7 sales before the app was hastily removed by Apple. Nonetheless, it inspired a bunch of copycats on the Android App Store, but if you search today, you’ll find all of these apps have disappeared as well. The high installation numbers are likely gamed by making the app was available for free at some point to get reviews and appear more legitimate.


Leaving this bad data in our dataset will misrepresent our analysis of the most expensive 'real' apps. 

### The most expensive apps sub $250

**Challenge:** Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

In [ ]:
df_apps_clean.drop(df_apps_clean[df_apps_clean["Price"] > 250].index, inplace=True)
# or: 
# df_apps_clean = df_apps_clean[df_apps_clean['Price'] < 250]
df_apps_clean.sort_values("Price", ascending=False).head()

App   Category  ...  Content_Rating     Genres
2281  Vargo Anesthesia Mega App    MEDICAL  ...        Everyone    Medical
1407               LTC AS Legal    MEDICAL  ...        Everyone    Medical
2629           I am Rich Person  LIFESTYLE  ...        Everyone  Lifestyle
2481    A Manual of Acupuncture    MEDICAL  ...        Everyone    Medical
2463         PTA Content Master    MEDICAL  ...        Everyone    Medical

[5 rows x 10 columns]

4 out of 5 of top paid apps are medical.

### Highest Grossing Paid Apps (ballpark estimate)

**Challenge:** Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?

In [ ]:
revenue_estimate = df_apps_clean.Price * df_apps_clean.Installs
df_apps_clean.insert(1, "Revenue_Estimate", revenue_estimate)

# or:
# df_apps_clean['Revenue_Estimate'] = df_apps_clean.Installs.mul(df_apps_clean.Price)

In [ ]:
df_apps_clean.sort_values("Revenue_Estimate", ascending=False).head(10)

App  ...                     Genres
9220                      Minecraft  ...  Arcade;Action & Adventure
8825                  Hitman Sniper  ...                     Action
7151  Grand Theft Auto: San Andreas  ...                     Action
7477            Facetune - For Free  ...                Photography
7977        Sleep as Android Unlock  ...                  Lifestyle
6594            DraStic DS Emulator  ...                     Action
6082                   Weather Live  ...                    Weather
7954                    Bloons TD 5  ...                   Strategy
7633        Five Nights at Freddy's  ...                     Action
6746     Card Wars - Adventure Time  ...    Card;Action & Adventure

[10 rows x 11 columns]

This generously assumes of course that all the installs would have been made at the listed price, which is unlikely, as there are always promotions and free give-aways on the App Stores.


The top spot of the highest-grossing paid app goes to … Minecraft at close to $70 million. It’s quite interesting that Minecraft (along with Bloons and Card Wars) is actually listed in the Family category rather than in the Game category. If we include these titles, we see that 7 out the top 10 highest-grossing apps are games. The Google Play Store seems to be quite flexible with its category labels.

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

If you were to release an app, would you choose to go after a competitive category with many other apps? Or would you target a popular category with a high number of downloads? Or perhaps you can target a category which is both popular but also one where the downloads are spread out among many different apps. That way, even if it’s more difficult to discover among all the other apps, your app has a better chance of getting installed, right? Let’s analyse this with bar charts and scatter plots and figure out which categories are dominating the market.

In [ ]:
df_apps_clean.Category.nunique()

33

In [ ]:
top10_category = df_apps_clean.Category.value_counts()[:10]
top10_category

FAMILY             1606
GAME                910
TOOLS               719
PRODUCTIVITY        301
PERSONALIZATION     298
LIFESTYLE           297
FINANCE             296
MEDICAL             292
PHOTOGRAPHY         263
BUSINESS            262
Name: Category, dtype: int64

### Vertical Bar Chart - Highest Competition (Number of Apps)

In [ ]:
bar = px.bar(x = top10_category.index,  # index = category name
             y = top10_category.values)

bar.show()

Based on the number of apps, the Family and Game categories are the most competitive. Releasing yet another app into these categories will make it hard to get noticed.

But what if we look at it from a different perspective? What matters is not just the total number of apps in the category but how often apps are downloaded in that category. This will give us an idea of how popular a category is. First, we have to group all our apps by category and sum the number of installations:

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

In [ ]:
category_installs = df_apps_clean.groupby("Category").agg({"Installs": pd.Series.sum})
category_installs.sort_values("Installs", ascending=True, inplace=True)

In [ ]:
h_bar = px.bar(x = category_installs.Installs,
               y = category_installs.index,
               orientation = "h",
               title = "Category Popularity")
 
h_bar.update_layout(xaxis_title='Number of Downloads', yaxis_title='Category')

h_bar.show()

Now we see that Games and Tools are actually the most popular categories. If we plot the popularity of a category next to the number of apps in that category we can get an idea of how concentrated a category is. Do few apps have most of the downloads or are the downloads spread out over many apps?

### Category Concentration - Downloads vs. Competition

**Challenge**: 
* First, create a DataFrame that has the number of apps in one column and the number of installs in another:

<img src=https://imgur.com/uQRSlXi.png width="350">


In [ ]:
cat_apps = df_apps_clean.groupby("Category").agg({"App": pd.Series.count})

In [ ]:
df_cat_merged = pd.merge(cat_apps, category_installs, on="Category")
print(f'The dimensions of the DataFrame are: {df_cat_merged.shape}')
df_cat_merged.head()

The dimensions of the DataFrame are: (33, 2)


App    Installs
Category                            
ART_AND_DESIGN        61   114233100
AUTO_AND_VEHICLES     73    53129800
BEAUTY                42    26916200
BOOKS_AND_REFERENCE  169  1665791655
BUSINESS             262   692018120

* Then use the [plotly express examples from the documentation](https://plotly.com/python/line-and-scatter/) alongside the [.scatter() API reference](https://plotly.com/python-api-reference/generated/plotly.express.scatter.html)to create scatter plot that looks like this. 

<img src=https://imgur.com/cHsqh6a.png>

*Hint*: Use the size, hover_name and color parameters in .scatter(). To scale the yaxis, call .update_layout() and specify that the yaxis should be on a log-scale like so: yaxis=dict(type='log') 

In [ ]:
scatter = px.scatter(df_cat_merged,  # data
                     x = "App",
                     y = "Installs",
                     title = "Category Concentration",
                     color = "Installs",
                     size = "App",
                     hover_name = df_cat_merged.index,
                     log_y = True,
                     range_x = [0, 1700])

scatter.update_layout(xaxis_title="Number of Apps (Lower=More Concentrated)",
                      yaxis_title="Installs")
                      # yaxis=dict(type="log"))  # or see above!


scatter.show()

What we see is that the categories like Family, Tools, and Game have many different apps sharing a high number of downloads. But for the categories like video players and entertainment, all the downloads are concentrated in very few apps.

# Extracting Nested Data from a Column

**Challenge**: How many different types of genres are there? Can an app belong to more than one genre? Check what happens when you use .value_counts() on a column with nested values? See if you can work around this problem by using the .split() function and the DataFrame's [.stack() method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html). 


In [ ]:
len(df_apps_clean.Genres.unique())

114

If we look at the number of unique values in the Genres column we get 114. But this is not accurate if we have nested data like we do here. We can see this using .value_counts() and looking at the values that just have a single entry. There we see that the semi-colon (;) separates the genre names.



In [ ]:
df_apps_clean.Genres.value_counts().sort_values(ascending=True)[:10]

Card;Brain Games                       1
Adventure;Brain Games                  1
Health & Fitness;Action & Adventure    1
Board;Pretend Play                     1
Role Playing;Brain Games               1
Communication;Creativity               1
Art & Design;Pretend Play              1
Racing;Pretend Play                    1
Arcade;Pretend Play                    1
Music & Audio;Music & Video            1
Name: Genres, dtype: int64

In [ ]:
stack = df_apps_clean.Genres.str.split(";", expand=True).stack()
print(f'We now have a single column with shape: {stack.shape}')
num_genres = stack.value_counts()
print(f'Number of genres: {len(num_genres)}')

We now have a single column with shape: (8564,)
Number of genres: 53
Tools                      719
Education                  587
Entertainment              498
Action                     304
Productivity               301
Lifestyle                  298
Personalization            298
Finance                    296
Medical                    292
Sports                     270
Photography                263
Business                   262
Communication              258
Health & Fitness           245
Casual                     216
News & Magazines           204
Social                     203
Simulation                 200
Travel & Local             187
Arcade                     185
Shopping                   180
Books & Reference          171
Video Players & Editors    150
Dating                     134
Puzzle                     124
Maps & Navigation          118
Role Playing               111
Racing                     103
Action & Adventure          96
Strategy                    95
F

# Colour Scales in Plotly Charts - Competition in Genres

**Challenge**: Can you create this chart with the Series containing the genre data? 

<img src=https://imgur.com/DbcoQli.png width=400>

Try experimenting with the built in colour scales in Plotly. You can find a full list [here](https://plotly.com/python/builtin-colorscales/). 

* Find a way to set the colour scale using the color_continuous_scale parameter. 
* Find a way to make the color axis disappear by using coloraxis_showscale. 

In [ ]:
bar_genre = px.bar(x = num_genres.index[:15],
                   y = num_genres.values[:15],
                   title="Top Genres",
                   color = num_genres.values[:15],
                   color_continuous_scale = "Aggrnyl")

bar_genre.update_layout(xaxis_title="Genre",
                        yaxis_title="Number of Apps",
                        coloraxis_showscale = False)

bar_genre.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

In [ ]:
df_apps_clean.Type.value_counts()

Free    7595
Paid     589
Name: Type, dtype: int64

We see that the majority of apps are free on the Google Play Store. But perhaps some categories have more paid apps than others. Let’s investigate. We can group our data first by Category and then by Type. Then we can add up the number of apps per each type. Using as_index=False we push all the data into columns rather than end up with our Categories as the index.



In [ ]:
df_free_vs_paid = df_apps_clean.groupby(["Category", "Type"], as_index=False).agg({"App": pd.Series.count})
df_free_vs_paid.sort_values("App")
# v few paid apps! Lots of free ones!

Category  Type   App
3    AUTO_AND_VEHICLES  Paid     1
24      FOOD_AND_DRINK  Paid     2
38  NEWS_AND_MAGAZINES  Paid     2
40           PARENTING  Paid     2
17       ENTERTAINMENT  Paid     2
..                 ...   ...   ...
31           LIFESTYLE  Free   284
21             FINANCE  Free   289
53               TOOLS  Free   656
25                GAME  Free   834
19              FAMILY  Free  1456

[61 rows x 3 columns]

**Challenge**: Use the plotly express bar [chart examples](https://plotly.com/python/bar-charts/#bar-chart-with-sorted-or-ordered-categories) and the [.bar() API reference](https://plotly.com/python-api-reference/generated/plotly.express.bar.html#plotly.express.bar) to create this bar chart: 

<img src=https://imgur.com/LE0XCxA.png>

You'll want to use the `df_free_vs_paid` DataFrame that you created above that has the total number of free and paid apps per category. 

See if you can figure out how to get the look above by changing the `categoryorder` to 'total descending' as outlined in the documentation here [here](https://plotly.com/python/categorical-axes/#automatically-sorting-categories-by-name-or-total-value). 

In [ ]:
bar_cat = px.bar(df_free_vs_paid,
                 x = "Category",
                 y = "App",
                 color = "Type",
                 title = "Free vs Paid Apps by Category",
                 log_y = True)

bar_cat.update_layout(xaxis_title="Category",
                      yaxis_title="Number of Apps",
                      barmode="group",
                      xaxis={'categoryorder':'total descending'})  # here

# bar_cat.update_xaxes(categoryorder="total descending")  # this or above

bar_cat.show()

What we see is that while there are very few paid apps on the Google Play Store, some categories have relatively more paid apps than others, including Personalization, Medical and Weather. So, depending on the category you are targeting, it might make sense to release a paid-for app.



But this leads to many more questions:

* How much should you charge? What are other apps charging in that category?

* How much revenue could you make?

* And how many downloads are you potentially giving up because your app is paid?

# Plotly Box Plots: Lost Downloads for Paid Apps

**Challenge**: Create a box plot that shows the number of Installs for free versus paid apps. How does the median number of installations compare? Is the difference large or small?

Use the [Box Plots Guide](https://plotly.com/python/box-plots/) and the [.box API reference](https://plotly.com/python-api-reference/generated/plotly.express.box.html) to create the following chart. 

<img src=https://imgur.com/uVsECT3.png>


In [ ]:
box = px.box(df_apps_clean,
             x = "Type",
             y = "Installs",
             log_y = True,
             color = "Type",
             title = "How Many Downloads are Paid Apps Giving Up?",
             notched = True,
             points="all")

box.update(layout_showlegend = False)

box.show()

From the hover text in the chart, we see that the median number of downloads for free apps is 500,000, while the median number of downloads for paid apps is around 5,000! This is massively lower.

But does this mean we should give up on selling a paid app? Let’s see how much revenue we would estimate per category.



# Plotly Box Plots: Revenue by App Category

**Challenge**: See if you can generate the chart below: 

<img src=https://imgur.com/v4CiNqX.png>

Looking at the hover text, how much does the median app earn in the Tools category? If developing an Android app costs $30,000 or thereabouts, does the average photography app recoup its development costs?

Hint: I've used 'min ascending' to sort the categories. 

In [ ]:
df_paid_apps = df_apps_clean[df_apps_clean["Type"] == "Paid"]

In [ ]:
box_cat = px.box(df_paid_apps,
                 x = "Category",
                 y = "Revenue_Estimate",
                 log_y = True,
                 title = "How Much Can Paid Apps Earn?")

box_cat.update_layout(xaxis_title="Category",
                      yaxis_title="Paid App Ballpark Revenue")

box_cat.update_xaxes(categoryorder="min ascending",
                     tickangle=30)

box_cat.update(layout_showlegend = False)

box_cat.show()

If an Android app costs $30,000 to develop, then the average app in very few categories would cover that development cost. The median paid photography app earned about 20,000 dollars. Many more app’s revenues were even lower - meaning they would need other sources of revenue like advertising or in-app purchases to make up for their development costs. However, certain app categories seem to contain a large number of outliers that have much higher (estimated) revenue - for example in Medical, Personalisation, Tools, Game, and Family.

So, if you were to list a paid app, how should you price it? To help you decide we can look at how your competitors in the same category price their apps.



# How Much Can You Charge? Examine Paid App Pricing Strategies by Category

**Challenge**: What is the median price price for a paid app? Then compare pricing by category by creating another box plot. But this time examine the prices (instead of the revenue estimates) of the paid apps. I recommend using `{categoryorder':'max descending'}` to sort the categories.

In [ ]:
df_paid_apps["Price"].median()

2.99

In [ ]:
box_price = px.box(df_paid_apps,
                   x = "Category",
                   y = "Price",
                   log_y = True,
                   title = "Price per Category")

box_price.update_xaxes(categoryorder="max descending",
                       tickangle=45)

box_price.show()

The median price for an Android app is $2.99.

However, some categories have higher median prices than others. This time we see that Medical apps have the most expensive apps as well as a median price of $5.49. In contrast, Personalisation apps are quite cheap on average at 1.49 dollars. Other categories which higher median prices are Business (4.99 dollars) and Dating (6.99 dollars). It seems like customers who shop in these categories are not so concerned about paying a bit extra for their apps.